numcosmo/lss/nc_galaxy_wl_ellipticity_kde.c

```diff
@@ -0,0 +1,367 @@
+  NcmVector *z_vec                             = nc_galaxy_redshift_spec_peek_z (NC_GALAXY_REDSHIFT_SPEC (gz));
+  gdouble min_g_i                              = GSL_POSINF;
+  gdouble max_g_i                              = GSL_NEGINF;
+  int j = 0;
+  gint gal_i;
+
+  ncm_stats_dist1d_free (self->kde);
+  ncm_vector_free (self->e_vec);
+
+  ncm_stats_dist1d_set_compute_cdf (sd1, FALSE);
+
+  for (gal_i = 0; gal_i < self->len; gal_i++)
+  {
+    const gdouble z_i = ncm_vector_get (z_vec, gal_i);
+    const gdouble r_i = ncm_matrix_get (wl_obs, gal_i, 0);
+    const gdouble g_i = ncm_matrix_get (wl_obs, gal_i, 1);
+    const gdouble s_i = nc_wl_surface_mass_density_reduced_shear (smd, dp, cosmo, r_i, z_i, z_cluster, z_cluster);
+
+    max_g_i = MAX (max_g_i, g_i);
+    min_g_i = MIN (min_g_i, g_i);
+    /*if ((g_i > 0.0 && g_i < 0.05 && s_i > 0.0 && s_i < 0.05))*/
+    //if (r_i > 0.07)
+    {
+      ncm_vector_set (g_vec, j, g_i);
+      ncm_stats_dist1d_epdf_add_obs (s_kde, s_i);
+      j++;
+    }
+  }
+
+  ncm_stats_dist1d_epdf_set_max (s_kde, max_g_i*1.01);
+  if (min_g_i > 0)
+    ncm_stats_dist1d_epdf_set_min (s_kde, min_g_i*0.99);
+  else
+    ncm_stats_dist1d_epdf_set_min (s_kde, min_g_i*1.01);
+  ncm_stats_dist1d_prepare (sd1);
+  {
+    const gdouble h = ncm_stats_dist1d_get_current_h (sd1);
+    const gdouble hp = gsl_hypot (h, ncm_matrix_get (wl_obs, 0, 2));
+
+    ncm_stats_dist1d_epdf_set_bw_type (s_kde, NCM_STATS_DIST1D_EPDF_BW_FIXED);
+    s_kde->h_fixed = hp;
+    ncm_stats_dist1d_prepare (sd1);
+  }
+
+  self->kde   = sd1;
+  self->e_vec = g_vec;
+
+// ncm_stats_dist1d_free (NCM_STATS_DIST1D (s_kde));
+// ncm_vector_free (g_vec);
+}
+
+// static void
+// _nc_galaxy_wl_ellipticity_kde_m2lnP_prep (NcGalaxyWLDist *gwld, NcHICosmo *cosmo, NcHaloDensityProfile *dp, NcWLSurfaceMassDensity *smd, const gdouble z_cluster, const guint gal_i)
+// {
+  
+// }
+
+static gdouble
+_nc_galaxy_wl_ellipticity_kde_m2lnP (NcGalaxyWLDist *gwld, NcHICosmo *cosmo, NcHaloDensityProfile *dp, NcWLSurfaceMassDensity *smd, const gdouble z_cluster, const guint gal_i, const gdouble z)
+{
+  NcGalaxyWLEllipticityKDE *gkde               = NC_GALAXY_WL_ELLIPTICITY_KDE (gwld);
+  NcGalaxyWLEllipticityKDEPrivate * const self = gkde->priv;
+  gdouble res                                  = 0.0;
+
+
+  const gdouble e_i = ncm_vector_get (self->e_vec, gal_i);
+  const gdouble p = ncm_stats_dist1d_eval_p (self->kde, e_i);
+  res += - 2 * log (p);
+  
+  return res;
+}
+
+static gdouble
+_nc_galaxy_wl_ellipticity_kde_gen (NcGalaxyWLDist *gwld, const gdouble g_true, NcmRNG *rng)
+{
+  NcGalaxyWLEllipticityKDE *grsg               = NC_GALAXY_WL_ELLIPTICITY_KDE (gwld);
+  NcGalaxyWLEllipticityKDEPrivate * const self = grsg->priv;
+  const gdouble sigma_g                        = ncm_matrix_get (self->obs, 0, 2);
+  
+  return ncm_rng_gaussian_gen (rng, g_true, sigma_g);
```

Comment:
Fix the sampling.

Reply:
I didn't know how to handle this. What is the purpose of this method?